synapse/federation/sender/per_destination_queue.py

```diff
@@ -580,3 +501,125 @@ def _start_catching_up(self) -> None:
         """
         self._catching_up = True
         self._pending_pdus = []
+
+
+@attr.s(slots=True)
+class _TransactionQueueManager:
+    """A helper async context manager for pulling stuff off the queues and
+    tracking what was last successfully sent, etc.
+    """
+
+    queue = attr.ib(type=PerDestinationQueue)
+
+    _device_stream_id = attr.ib(type=Optional[int], default=None)
+    _device_list_id = attr.ib(type=Optional[int], default=None)
+    _last_stream_ordering = attr.ib(type=Optional[int], default=None)
+
+    async def __aenter__(self) -> Tuple[List[EventBase], List[Edu]]:
+        # We have to keep 2 free slots for presence and rr_edus
+        limit = MAX_EDUS_PER_TRANSACTION - 2
+
+        device_update_edus, dev_list_id = await self.queue._get_device_update_edus(
+            limit
+        )
+
+        if device_update_edus:
+            self._device_list_id = dev_list_id
+        else:
+            self.queue._last_device_list_stream_id = dev_list_id
```

Comment:
It looks like we used to do this inside of the `if not pending_pdus and not pending_edus` block. Is this a behavior change?

Reply:
No, we just need to only set it after we've sent all device EDUs that were returned, and if none are returned we can immediately set it. If we don't set it when we get none returned then that is also fine, just slightly less efficient (as the DB query needs to look at a larger range).